datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -289,6 +289,14 @@ impl PhysicalExpr for BinaryExpr {
             return apply_cmp_for_nested(self.op, &lhs, &rhs);
         }
 
+        if left_data_type.is_floating() {
+            lhs = normalize_floating_zeros(lhs, &left_data_type)?;
+        };
+
+        if right_data_type.is_floating() {
+            rhs = normalize_floating_zeros(rhs, &right_data_type)?;
+        }
```

Comment:
I'm curious to know how much overhead we will be adding to all binary operations on floating point numbers with this normalization approach. It would be good to have some criterion benchmarks in place so we can assess this.

The alternative to normalizing zeroes would be to update the comparison kernels to handle the negative zero case and this would avoid introducing extra array transformations.

Reply:
Also, given that this normalization function is iterative, it may be worth adding a check that the array contains a negative zero before calling.

Reply:
I agree with @andygrove -- this should be taken care of by the kernel. This could turn into a non-negligible overhead that could have been avoided.

Reply:
Initial criterion results on 1,000,000 rows

```
evaluate with normalization
                        time:   [20.411 ms 20.570 ms 20.747 ms]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

evaluate without normalization
                        time:   [820.15 µs 825.46 µs 831.99 µs]
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
```

That's almost 200% worse.

I'll try a couple changes first, if the performance is still horrible I'll close the PR.

Reply:
If the check overhead looks ok, I can try to reduce the 20-25x overhead to match the 3x check overhead.

Reply:
Let's just write a `numeric_eq` kernel and use it instead.

Reply:
Following up on this: Is the proposal here to add a new kernel to arrow-rs that supports these floating point comparison semantics, or hoist arrow-rs comparison code to DataFusion and implement it there?

Reply:
I think we can start with the latter, make sure we get the impl right (no performance regressions etc.) and then discuss moving it upstream to arrow-rs